contracts/src/v0.8/functions/dev/1_0_0/FunctionsBilling.sol

```diff
@@ -0,0 +1,421 @@
+    RequestBilling memory billing = RequestBilling(subscriptionId, msg.sender, gasLimit, gasPrice);
+    uint96 donFee = getDONFee(data, billing);
+    uint96 adminFee = getAdminFee(data, billing);
+    return _calculateCostEstimate(gasLimit, gasPrice, donFee, adminFee);
+  }
+
+  /**
+   * @notice Uses current price feed data to estimate a cost
+   */
+  function _calculateCostEstimate(
+    uint32 gasLimit,
+    uint256 gasPrice,
+    uint96 donFee,
+    uint96 adminFee
+  ) internal view returns (uint96) {
+    int256 weiPerUnitLink;
+    weiPerUnitLink = getFeedData();
+    if (weiPerUnitLink <= 0) {
+      revert InvalidLinkWeiPrice(weiPerUnitLink);
+    }
+    uint256 executionGas = s_config.gasOverhead + s_config.gasAfterPaymentCalculation + gasLimit;
+    // (1e18 juels/link) (wei/gas * gas) / (wei/link) = juels
+    uint256 paymentNoFee = (1e18 * gasPrice * executionGas) / uint256(weiPerUnitLink);
+    uint256 fee = uint256(donFee) + uint256(adminFee);
+    if (paymentNoFee > (1e27 - fee)) {
+      revert PaymentTooLarge(); // Payment + fee cannot be more than all of the link in existence.
+    }
+    return uint96(paymentNoFee + fee);
+  }
+
+  // ================================================================
+  // |                       Billing Methods                        |
+  // ================================================================
+  /**
+   * @notice Initiate the billing process for an Functions request
+   * @param data Encoded Chainlink Functions request data, use FunctionsClient API to encode a request
+   * @param billing Billing configuration for the request
+   * @return requestId - A unique identifier of the request. Can be used to match a request to a response in fulfillRequest.
+   * @return estimatedCost
+   * @return gasAfterPaymentCalculation
+   * @return requestTimeoutSeconds
+   * @dev Only callable by the Functions Router
+   */
+  function _startBilling(
+    bytes memory data,
+    RequestBilling memory billing
+  ) internal returns (bytes32, uint96, uint256, uint256) {
+    // No lower bound on the requested gas limit. A user could request 0
+    // and they would simply be billed for the gas and computation.
+    if (billing.gasLimit > s_config.maxGasLimit) {
+      revert GasLimitTooBig(billing.gasLimit, s_config.maxGasLimit);
+    }
+
+    // Check that subscription can afford the estimated cost
+    uint96 donFee = getDONFee(data, billing);
+    uint96 adminFee = getAdminFee(data, billing);
+    uint96 estimatedCost = _calculateCostEstimate(billing.gasLimit, billing.gasPrice, donFee, adminFee);
+    IFunctionsSubscriptions subscriptions = IFunctionsSubscriptions(address(s_router));
+    (uint96 balance, uint96 blockedBalance, , , ) = subscriptions.getSubscription(billing.subscriptionId);
+    (, uint64 initiatedRequests, ) = subscriptions.getConsumer(billing.client, billing.subscriptionId);
+
+    if (balance - blockedBalance < estimatedCost) {
+      revert InsufficientBalance();
+    }
```

Comment:
Where do we update the new blockedBalance?

Reply:
In the Router